python/kwiver/sprokit/adapters/adapter_data_set.cxx

```diff
@@ -112,30 +112,30 @@ py::object get_port_data_correct_type(adapter_data_set &self, ::sprokit::process
 {
   kwiver::vital::any const any = self.get_port_data<kwiver::vital::any>(port);
 
-  #define GET_OBJECT(TYPE) \
-  if (typeid(TYPE).hash_code() == any.type().hash_code()) \
+  #define ADS_GET_OBJECT(TYPE) \
+  if (any.is_type<TYPE>()) \
```

Comment:
I don't understand the C++ here enough to know what is happening. Can I get an explanation of old versus new behavior?

Reply:
So in an ideal world, the implementation with hash codes would just work. However, under lib C++, this comparison will fail if two equivalent types are defined in different shared libraries. This is what was causing the Mac errors. Is_type compares the names of the type. Using is_type is probably the best option regardless, since it's in vital.

Reply:
I see, `vital.any.type()` and `typeid(.)` return a `std::type_info`, which contains the `hash_code` method.   

When you say two libs define equivalent types, you mean that two libs share type definitions that have the same names (and maybe the same methods)? In that case I can understand why the `hash_code` would be different. Its like in Python if you defined the same class in two files, using isinstance(foo.Bar(), baz.Bar) is going to (and should) return False even if Bar is functionally the same. 

It seems slightly dangerous to only rely on the names here given that its very easy for another lib to define a class with the same name, but I guess that's the behavior we want?  

It might be worth adding a comment discussing this behavior in case this comes back to bite us in the future.

Reply:
> When you say two libs define equivalent types, you mean that two libs share type definitions that have the same names (and maybe the same methods)?

Not quite, sorry for not being clear. When 2 pybind modules have bindings for the same C++ type, _that_ `type_info`
/ `type_info.hash_code()`  comparison fails. In our example, the pybind modules `datum` and `adapter_data_set` both provide bindings for `std::vector<double>` (and the other vector instantiations). Consider the following code:

`> ads = AdapterDataSet.create()`
` # Notice that we use datum.VectorDouble(), not adapter_data_set.VectorDouble()`
`> d = datum.new(datum.VectorDouble([5.2, -3.14]))`
`> ads["p1"] = d`
`> ads["p1"]`

That last call to `get_port_data_correct_type` is what would cause issues. In that function, we eventually check if the object at that port is an `std::vector<double>`, which in this case, should return true. The issue is that the `std::vector<double>` stored at that port was created from the `datum` module, not the `adapter_data_set` module. The `std::type_info`'s and their `hash_codes()` are not the same, even though the underlying C++ type is the same, so the comparison would return false.

If you'd still prefer that I add a comment, I'm happy to do so.

Reply:
I think that is roughly what I was thinking. Two pybind modules would effectively be different C++ shared libraries (i.e. different .so files). In each of those modules you create a new python class VectorDouble that has the same underlying C++ implementation, but because it was defined separately in each pybind module, it gets a different type hash code. I think its a non-trivial enough issue that a comment pointing back to this conversation is warranted.

Reply:
Added a comment in b2ca3d8828aa87126cb981b57a759eda20e25002.

We are comparing the underlying C++ types, not the Python types/names.